### PR TITLE
Allow empty YAML fields

### PR DESCRIPTION
**Pull request by iojw**
Users may often comment out the nested contents of YAML fields without remove the top-level field. Previously, this would throw the following error due to our YAML validation:
```
ValueError: Invalid task YAML: None is not of type 'string'. Check problematic field(s): $.run
```

This PR updates it so that such fields are treated as though they do not exist. It does this by removing all fields with a `None` value in the parsed dictionary before validation. 

We do not solve this issue by adding `null` to the JSON schema for our YAML config because many places in the code test for the existence of specific fields in the dictionary so we would have to modify all of their logic. Moreover, having to add `null` to every single field in the JSON schema reduces readability.

Tested (run the relevant ones):

- [x] Tested with YAML configs with empty fields
- [x] Added unit test for yaml parsing 
- [ ] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `bash tests/backward_comaptibility_tests.sh`

### CODE REVIEW

File: sky/data/storage.py [hunk 1]
```diff
@@ -766,6 +766,7 @@ def warn_for_git_dir(source: str):
 
     @classmethod
     def from_yaml_config(cls, config: Dict[str, Any]) -> 'Storage':
+        config = {k: v for k, v in config.items() if v is not None}
```
Comment (concretevitamin) [inline sky/data/storage.py:769]: Possible to move this into backend_utils.validate_schema since there are three places?

**Review by concretevitamin — COMMENTED** [review]
Thanks @iojw. A nit. Consider adding a unit test to ensure this doesn’t break in the future?

**Comment by concretevitamin** [comment]
Have we ensured that for all fields in the current YAML spec, None means absence means default value? Any cases where absence means something else?

File: sky/data/storage.py (hunk 1 above)
Comment (iojw) [reply sky/data/storage.py:769]: Possible, but that would remove the possibility of us having nullable YAML fields in the future. I'll add a flag so we can decide whether or not we want to remove None values.

Comment (iojw) [reply sky/data/storage.py]: I've added a flag so we can enable this for validation. The reason we don't want to do it for all cases is that it prevents us from actually having null values in the schema in the future.

In addition, a lot of the current code relies on the fact that if a key exists in the config dictionary, then the value is valid. E.g. if we allow None to be inside the config dictionary, then inside `storage.from_yaml_config`, `force_delete = config.pop('_force_delete', False)` would incorrectly set `force_delete` to None if the user specifies `force_delete` in the YAML without a value. Therefore, I also had to refactor the code such that it properly handle `None` config values.

**Comment by iojw** [comment]
@concretevitamin Yes, I've looked through `Resources`, `Storage`, and `Task` to ensure that for these schemas, `None` does not have a special meaning and is properly handled.

File: sky/backends/backend_utils.py [hunk 2]
```diff
@@ -2471,7 +2471,10 @@ def stop_handler(signum, frame):
         raise KeyboardInterrupt(exceptions.SIGTSTP_CODE)
 
 
-def validate_schema(obj, schema, err_msg_prefix=''):
+def validate_schema(obj, schema, err_msg_prefix='', skip_none=False):
```
Comment (concretevitamin) [inline sky/backends/backend_utils.py:2474]: Let's add a docstr, including Args:?

File: sky/backends/onprem_utils.py [hunk 3]
```diff
@@ -488,7 +488,8 @@ def check_local_cloud_args(cloud: Optional[str] = None,
         yaml_config: User's task yaml loaded into a JSON dictionary.
     """
     yaml_cloud = None
-    if yaml_config is not None and 'resources' in yaml_config:
+    if yaml_config is not None and 'resources' in yaml_config and yaml_config[
+            'resources'] is not None:
```
Comment (concretevitamin) [inline sky/backends/onprem_utils.py:492]:
```suggestion
    if yaml_config is not None and yaml_config.get('resources', None) is not None:
```

File: tests/test_yaml_parser.py
Comment (concretevitamin) [inline tests/test_yaml_parser.py:1]: Add to: .github/workflows

File: tests/test_yaml_parser.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
+import pathlib
+import textwrap
+
+import pytest
+from sky.task import Task
```
Comment (concretevitamin) [inline tests/test_yaml_parser.py:5]:
```suggestion
import pytest

from sky.task import Task
```

File: tests/test_yaml_parser.py [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+def test_empty_fields_task(tmp_path):
+    config_path = _create_config_file(
+        textwrap.dedent(f"""\
+            name: task
+            resources:
+            workdir: examples/
+            file_mounts:
+            setup: echo "Running setup."
+            num_nodes:
+            run:
```
Comment (concretevitamin) [inline tests/test_yaml_parser.py:23]:
```suggestion
            run:
              # commented out, empty run
```

File: tests/test_yaml_parser.py [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+            """), tmp_path)
+    task = Task.from_yaml(config_path)
+
+    resources = list(task.resources)[0]
+    assert resources.cloud is None
+    assert resources.region is None
+    assert resources.accelerators == {'V100': 1}
+    assert resources.disk_size is 256
+    assert resources.use_spot is False
+    assert resources.cpus == "32"
```
Comment (concretevitamin) [inline tests/test_yaml_parser.py:67]:
```suggestion
    assert resources.cpus == '32'
```

File: tests/test_yaml_parser.py [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+def test_invalid_fields_task(tmp_path):
+    config_path = _create_config_file(
+        textwrap.dedent(f"""\
+            name: task
+
+            not_a_valid_field:
+            """), tmp_path)
+    with pytest.raises(AssertionError) as e:
+        Task.from_yaml(config_path)
+    assert 'Invalid task args' in e.value.args[0]
```
Comment (concretevitamin) [inline tests/test_yaml_parser.py:45]: indent inconsistent w/ L79?

File: sky/backends/backend_utils.py (hunk 2 above)
Comment (concretevitamin) [inline sky/backends/backend_utils.py:2474]: When do we want skip_none=False? In cli.py’s admin_deploy()?

File: sky/resources.py [hunk 8]
```diff
@@ -867,40 +867,40 @@ def from_yaml_config(cls, config: Optional[Dict[str, str]]) -> 'Resources':
...
-            resources_fields['instance_type'] = config.pop('instance_type')
-        if config.get('cpus') is not None:
-            resources_fields['cpus'] = str(config.pop('cpus'))
-        if config.get('memory') is not None:
-            resources_fields['memory'] = str(config.pop('memory'))
-        if config.get('accelerators') is not None:
-            resources_fields['accelerators'] = config.pop('accelerators')
-        if config.get('accelerator_args') is not None:
+        resources_fields['cloud'] = clouds.CLOUD_REGISTRY.from_str(
+            config.pop('cloud', None))
```
Comment (concretevitamin) [inline sky/resources.py:877]: These changes may be harmless but change previous semantics. E.g., previously, `resources_fields['cloud']` doesn't get assigned if `cloud` field is not in `config`; but now it does.

Is it cleaner to revert the changes here somehow? Maybe make `validate_schema` remove None values? Note: this is just a question for discussion.

**Review by concretevitamin — COMMENTED** [review]
Thanks, some questions.

File: tests/test_yaml_parser.py (hunk 7 above)
Comment (iojw) [reply tests/test_yaml_parser.py:45]: You're right, thanks for catching this!

File: sky/resources.py (hunk 8 above)
Comment (iojw) [reply sky/resources.py:877]: Agreed that it is changes semantics. The reason why these changes are required is because the previous code had an implicit assumption that there are no `None` values in the code, so it depends if this something we want to avoid. Personally, I think that it is better to avoid that assumption and handle `None` explicitly, even if this may lead to semantic changes.

File: sky/backends/backend_utils.py (hunk 2 above)
Comment (iojw) [reply sky/backends/backend_utils.py]: We may want it in cases where `None` is a valid value for some config we are validating. After thinking about it further, it makes more sense to set the default to `True`, since it is likelier than we want to remove `None` values than not. I've updated to code for this.

**Comment by concretevitamin** [comment]
@iojw Oops, slipped through the radar. Could you merge with master first?

**Comment by iojw** [comment]
@concretevitamin Updated!

**Comment by concretevitamin** [comment]
LGTM. With the latest commit, 
```
FAILED tests/test_smoke.py::test_gcp_stale_job_manual_restart - AttributeError: module 'sky.skylet.events' has no attribute 'JobUpdateEvent'. Did you mean: 'SpotJobUpdateEvent'?
FAILED tests/test_smoke.py::test_file_mounts - Exception: test failed: less /tmp/using_file_mounts-bsvqi8s_.log
FAILED tests/test_smoke.py::test_large_job_queue - Exception: test failed: less /tmp/large_job_queue-x96b5ene.log
FAILED tests/test_smoke.py::test_job_queue_multinode - Exception: test failed: less /tmp/job_queue_multinode-iyo5bv9n.log
FAILED tests/test_smoke.py::test_multi_echo - Exception: test failed: less /tmp/multi_echo-86fyim3q.log
FAILED tests/test_smoke.py::test_env_check - Exception: test failed: less /tmp/env_check-v7b2xnyx.log
6 failed, 62 passed, 35 skipped, 1646 warnings in 3462.05s (0:57:42)
```
These tests failed, most likely due to merge conflicts. See comment: https://github.com/skypilot-org/skypilot/pull/1636/files#r1220710715

Could you merge again?

**Review by concretevitamin — APPROVED** [review]
LGTM, thanks @iojw.

**Comment by iojw** [comment]
Confirmed that smoke tests pass with these changes.
